sdk/schemaregistry/azure-data-schemaregistry-apacheavro/src/main/java/com/azure/data/schemaregistry/apacheavro/SchemaRegistrySchemaCache.java

```diff
@@ -0,0 +1,222 @@
+ */
+class SchemaRegistrySchemaCache {
+    private static final String SIZE_KEY = "size";
+    private static final String TOTAL_LENGTH_KEY = "totalLength";
+
+    private final ClientLogger logger = new ClientLogger(SchemaRegistrySchemaCache.class);
+
+    private final SchemaCache cache;
+    private final SchemaRegistryAsyncClient schemaRegistryClient;
+    private final String schemaGroup;
+    private final boolean autoRegisterSchemas;
+    private final Object lock = new Object();
+
+    SchemaRegistrySchemaCache(SchemaRegistryAsyncClient schemaRegistryClient, String schemaGroup,
+        boolean autoRegisterSchemas, int capacity) {
+
+        this.schemaRegistryClient = schemaRegistryClient;
+        this.schemaGroup = schemaGroup;
+        this.autoRegisterSchemas = autoRegisterSchemas;
+        this.cache = new SchemaCache(capacity);
+    }
+
+    Mono<String> getSchemaId(Schema schema) {
+        final String existingSchemaId;
+        synchronized (lock) {
+            existingSchemaId = cache.getSchemaId(schema);
+        }
+
+        if (existingSchemaId != null) {
+            return Mono.just(existingSchemaId);
+        }
+
+        final String schemaFullName = schema.getFullName();
+        final String schemaString = schema.toString();
+
+        // It is possible to create the serializer without setting the schema group. This is the case when
+        // autoRegisterSchemas is false. (ie. You are only using it to deserialize messages.)
+        if (CoreUtils.isNullOrEmpty(schemaGroup)) {
+            return monoError(logger, new IllegalStateException("Cannot serialize when 'schemaGroup' is not set. Please"
+                + "set in SchemaRegistryApacheAvroSerializer.schemaGroup when creating serializer."));
+        }
+
+        final Mono<SchemaProperties> serviceCall;
+        if (autoRegisterSchemas) {
+            serviceCall = this.schemaRegistryClient
+                .registerSchema(schemaGroup, schemaFullName, schemaString, SchemaFormat.AVRO);
+        } else {
+            serviceCall = this.schemaRegistryClient.getSchemaProperties(
+                schemaGroup, schemaFullName, schemaString, SchemaFormat.AVRO);
+        }
+
+        return serviceCall.map(properties -> {
+            final String schemaId = properties.getId();
+
+            synchronized (lock) {
+                cache.put(schemaId, schema);
+                logCacheStatus();
+            }
+
+            return schemaId;
+        });
+    }
+
+    Mono<Schema> getSchema(String schemaId) {
+        synchronized (lock) {
+            final Schema existing = cache.get(schemaId);
+
+            if (existing != null) {
+                return Mono.just(existing);
+            }
+        }
+
+        return schemaRegistryClient.getSchema(schemaId)
+            .handle((registryObject, sink) -> {
+                final String schemaString = registryObject.getDefinition();
+
+                final Schema parsedSchema;
+                synchronized (lock) {
```

Comment:
Have we perf tested the impact of locking in Event Hubs scenario where every event will do a getSchema for deserializing the event?

Reply:
I think this is required to keep it thread-safe. My guidance around this was from the Javadocs:

> Note that this implementation is not synchronized. If multiple threads access a linked hash map concurrently, and at least one of the threads modifies the map structurally, it must be synchronized externally. This is typically accomplished by synchronizing on some object that naturally encapsulates the map. If no such object exists, the map should be "wrapped" using the [Collections.synchronizedMap](https://docs.oracle.com/javase/8/docs/api/java/util/Collections.html#synchronizedMap-java.util.Map-) method. This is best done at creation time, to prevent accidental unsynchronized access to the map: